### ISSUE TITLE
arrayToHtmlList update

### ISSUE THREAD

**Issue by skatcat31** [30-seconds/30-seconds-of-code#829]
## Description

So our function `arrayToHtmlList` is in a pretty bad space and could use an update. We already had some discussion over at #675 but I wanted to move the update discussion over to an issue for more visibility.

Current problem:

It uses an awkward query selector instead of the `document.prototype.getElementByID()` function, and it uses an IIFE to declare an intermediate variable closure when we could do with just API chaining the return values.

Another issue is we are just forcing it inside a list item. We could easily allow a formatter to be passed as the last argument with the map footprint and a default that won't change functionality, however the change in footprint because a bit... awkward as it becomes `data`, `binding`, `callback`, but it was an old monster to begin with.

With those two issues in mind we should look at either updating or cleaning up the snippet so it is more maintainable than it's current form or decide to leave as it.

**Comment by Siarhei-Zharnasek** [comment]
```
const arrayToHtmlList = (arr, selector, elemTag) =>
  (
    container => container.innerHTML += arr.reduce((acc, el) => `${acc}<${elemTag}>${el}</${elemTag}>`, '')
  )(document.querySelector(selector));
```

@skatcat31 what do you think about implementing this way?

**Comment by skatcat31** [comment]
@Siarhei-Zharnasek I feel like if we're going to do that we should do a formatter function so they can choose their own entire string

**Comment by Chalarangelo** [comment]
- +1 on `document.selectElementById`. Either that or we should pass a query selector and use `document.querySelector` instead. Either way, it's awkward right now.
- IIFE has to go, let's do something better instead.
- I like the formatter idea. From a usability perspective, it's a lot better. From a learning standpoint, this might land the snippet in advanced (depending on verbosity), but it's a very good teaching opportunity.

TL;DR: Let's fix that, it's an awkward snippet and all the proposed suggestions are pretty good.

**Comment by Ankitb09** [comment]
`function arrayToHtmlList(arr, id, elemTag) {
    let parentElem = document.getElementById(id);
    parentElem.innerHTML += arr.map((item) => {
        return `<${elemTag}>${item}</${elemTag}>`
    }).join('')
}`

@skatcat31 What do you think about this?

**Comment by skatcat31** [comment]
@Ankitb09 again what advantage does that bring over adding a default formatter so they can handle the full string if wanted?

**Comment by konstantin-hatvan** [comment]
Hello,

I have never contributed to open source before but this issue seemed like a good one to start. I hope I can contribute.

```javascript
const arrayToHtmlList = (arr, listID, formatterString = '<li></li>', marker = '|') => {
  document.querySelector(`#${listId}`).innerHTML += arr.map(item => formatterString.replace(marker, item)).join('');
});
```

What do you think about this?

**Comment by konstantin-hatvan** [comment]
Sorry i made a typo in the above code ...  The marker was missing in the formatterString variableShould be:

```javascript
const arrayToHtmlList = (arr, listID, formatterString = '<li>|</li>', marker = '|') => {
  document.querySelector(`#${listId}`).innerHTML += arr.map(item => formatterString.replace(marker, item)).join('');
});
```

An example usage of this would be

```javascript
arrayToHtmlList(['item 1', 'item 2'], 'listId'); // use default template string
arrayToHtmlList(['item 1', 'item 2'], 'listId', '<li class="test">This is the added item: |</li>'); // only use default replace marker but overwrite template string
arrayToHtmlList(['item 1', 'item 2'], 'listId', '<li class="test">This is the added item: ###</li>', '###'); // overwrite template string and replace marker --> full flexibility
```

**Comment by skatcat31** [comment]
You're right I was missing a closing arguments parens.

Reduce might be an option

```js
const arrayToHtmlList = (arr, listID, formatterFunction = (acc, item) => ((acc +=`<li>${item}</li>`), acc)) => {
  document.getElementById(listID).innerHTML = arr.reduce(formatterFunction, '');
};
```

however we run into a lot of really nasty parens in the footprint and having to keep around an empty string in body, which is something we need to do regardless(the join). It does however reduce the time complexity by a significant margin(O(n) vs (2n))

This however means that any formatter passed will be a reduce function and thus must remember the accumulator, or in other words more developer debt versus operation debt.

So then the question becomes do we allow more time complexity or more developer complexity?

We could avoid this by using a custom for loop accumulator so that a map compatible function gets passed, manually applied, and then added to the base string, solving both problems.

**Comment by konstantin-hatvan** [comment]
@skatcat31 Although this is not the most aesthetic implementation, it's the best solution in my eyes.

I think it is preferable to overwrite the innerHTML rather than concatenating because of the cleaner representation, as you have mentioned before.

**Comment by Chalarangelo** [comment]
@skatcat31 Wanna rewrite this one as discussed? Or I could give it a shot in a few days.

**Comment by nickforddesign** [comment]
I’d recommend making this a pure function that just returns html, leave the DOM out of it.

**Comment by skatcat31** [comment]
That's probably for the best. It means we'd be releasing a breaking change, but to be fair this thing is in a bad spot. If we REALLY wanted to we could actually returns a full nodelist so that other modifications could be done, but I think at that point a stand alone script library would be better

**Comment by skatcat31** [comment]
so yeah I'd be okay with it just returning an HTML string instead

**Comment by freality** [comment]
Hey all!

Just found (and really like) your project! Great work.

I'd like to through an implementation into the ring. Essentially a modification of a few previous suggestions.
```javascript
const arrayToHtmlList = (arr, listID, formatterFn = (item) => `<li>${item}</li>`) => {
    document.getElementById(listID).innerHTML += arr.reduce(
        (acc, curr, idx) => acc += formatterFn(curr, idx)
        , ''
    );
};
```

Also, it strikes me that by adding the default formatterFunction, this function effectively becomes `appendArrayOfItemsToElement`.

Example:
```javascript
arrayToHtmlList(['foo.js', 'bar.js', 'baz.js'], 'body-id', (src) => `<script type="text/javascript" src="${src}"></script>`);
```

**Comment by skatcat31** [comment]
Man this shows how busy a full time high cognitive load job gets. We should make a final decision on this soon and release the change

**Comment by samuelcolburn** [comment]
2 more suggestions:

1. The same as @skatcat31 's most recent `Array.reduce` method but using insertAdjacentHTML instead of innerHTML:
```javascript
const arrayToHtmlList = (arr, listID, formatterFunction = (acc, item) => ((acc +=`<li>${item}</li>`), acc)) =>
  document
    .getElementById(listID)
    .insertAdjacentHTML('beforeend', arr.reduce(formatterFunction, ''));
```
Reasoning:

> The insertAdjacentHTML() method of the Element interface parses the specified text as HTML or XML and inserts the resulting nodes into the DOM tree at a specified position. It does not reparse the element it is being used on, and thus it does not corrupt the existing elements inside that element. This avoids the extra step of serialization, making it much faster than direct innerHTML manipulation.

Source: [MDN Element.insertAdjacentHTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentHTML)


2. Similar to @nickforddesign 's suggestion, creates and returns new element instead of manipulating an existing element:
```javascript
const arrayToHtmlList = (
  arr,
  tagName = 'ul',
  formatterFunction = (acc, item) => ((acc +=`<li>${item}</li>`), acc)
) => {
  const list = document.createElement(tagName);
  list.insertAdjacentHTML('beforeend', arr.reduce(formatterFunction, ''));
  return list;
};
```

**Comment by jasonralston100** [comment]
> const arrayToHtmlList = (arr, listID, formatterFunction = (item) => `<li>${item}</li>`) => {
>   document.getElementById(listID).innerHTML = arr.map(formatterFunction).join('');
> };

This is the easiest and cleanest way to take a constant array and convert it to an array list with it's inner.HTML. If it were PowerShell I would just use a HTML fragment for the formatting. Not sure what your gaining here. It should output just fine or you could convert to straight xml? this is just output `<li>${item}</li>`

**Comment by vkv88** [comment]
Hi   I  would like to   add something**:**
**.**using a  classic  for loop  will be  more efficient   than  a  reduce function  especially when converting big array .
by the way this is my first  contribution on GitHub👶 
also   i'm writing   a new  snippet right now   any suggestion other than what's mentioned in the contribution Guideline

**Comment by meysam81** [comment]
After reading through all the comments and the previous [PR](https://github.com/30-seconds/30-seconds-of-code/issues/675), I would also suggest returning just the HTML and letting the developer decide what to do with it. As the name also suggests "array to **html list**". It does not imply anything about changing the DOM.

peace :v:

**Comment by Chalarangelo** [comment]
@meysam81 I think we will probably stick with that and maybe add a blog for clarification/exploration on the subject.

**Comment by Chalarangelo** [comment]
I'm closing this and will probably add a blog soon to clarify and explore ideas from this thread.
